File: runners/php/verify_wba.php
<?php
// PHP runner for the Web Bot Auth profile corpus (webbotauth_v0).
//
// Self-contained PHP port of the Python reference runner
// (runners/python/verify_wba.py) and its independent KAT gate
// (tools/check_kat_wba.py). Every PROFILE rule (signing base, coverage
// completeness, freshness window, directory keyid resolution, Signature-Agent
// SSRF, and the Ed25519 signature verdict) is re-implemented here directly from
// the corpus `policy` block, NOT imported from a shared oracle, so a runner
// passing is genuine agreement rather than an echo. Ed25519 is verified with
// ext-sodium (sodium_crypto_sign_verify_detached) over the raw 32-byte key; the
// IPv4/IPv6 CIDR membership needed by the SSRF guard is hand-rolled with
// inet_pton + bitmask (PHP has no built-in CIDR support).
//
// Mirrors runners/python/verify_wba.py, runners/typescript/verify_wba.mjs and
// runners/go/verify_wba_go.go case for case.
//
//   php -d extension=sodium -d extension=gmp runners/php/verify_wba.php [corpus.json]
//
// Corpus path: argv[1], else ../../corpus/webbotauth_v0/webbotauth_v0.json
// relative to this runner. Exit 0 iff every case matches, else exit 1.

class SigningBaseError extends Exception {}

const DEFAULT_PATH = __DIR__ . "/../../corpus/webbotauth_v0/webbotauth_v0.json";

// ---------------------------------------------------------------------------
// Rule 1: RFC 9421 Section 2.5 signing base, by direct concatenation.
// Throws when a covered component's value is missing (build impossible).
// ---------------------------------------------------------------------------
function build_signing_base(array $in, ?string $params_raw): string {
    $lines = [];
    foreach ($in["covered_components"] as $component) {
        $name = strtolower((string)$component);
        switch ($name) {
            case "@authority":
                $val = $in["authority"] ?? null;
                break;
            case "@method":
                $val = $in["method"] ?? null;
                break;
            case "@path":
                $val = $in["path"] ?? null;
                break;
            default:
                $headers = $in["headers"] ?? [];
                $val = array_key_exists($name, $headers) ? $headers[$name] : null;
        }
        if ($val === null) throw new SigningBaseError("missing covered component: $name");
        $lines[] = "\"$name\": $val";
    }
    $lines[] = "\"@signature-params\": " . (string)$params_raw;
    return implode("\n", $lines);
}

// ---------------------------------------------------------------------------
// Rule 2: coverage completeness (every required component present, lowercased).
// ---------------------------------------------------------------------------
function coverage_ok(array $covered, array $policy): bool {
    $have = [];
    foreach ($covered as $c) $have[strtolower((string)$c)] = true;
    foreach ($policy["required_covered_components"] as $req) {
        if (!isset($have[strtolower((string)$req)])) return false;
    }
    return true;
}

// ---------------------------------------------------------------------------
// Rule 3: freshness window.  All three must be integers, expires > created,
// and (created - skew) <= now < expires.
// ---------------------------------------------------------------------------
function freshness_ok($created, $expires, $now, array $policy): bool {
    $skew = $policy["clock_skew_seconds"] ?? 0;
    if (!is_int($created) || !is_int($expires) || !is_int($now)) return false;
    if ($expires <= $created) return false;
    return ($created - $skew) <= $now && $now < $expires;
}

// ---------------------------------------------------------------------------
// Rule 4: directory keyid resolution to a usable Ed25519 JWK.
// ---------------------------------------------------------------------------
function directory_ok(array $directory, string $keyid): bool {
    foreach ($directory as $jwk) {
        if (($jwk["kid"] ?? null) === $keyid) {
            return ($jwk["kty"] ?? null) === "OKP"
                && ($jwk["crv"] ?? null) === "Ed25519"
                && (string)($jwk["x"] ?? "") !== "";
        }
    }
    return false;
}

// ---------------------------------------------------------------------------
// Rule 5: Signature-Agent SSRF guard.  Needs IPv4 AND IPv6 CIDR membership by
// bit masking on the inet_pton-packed bytes (PHP has no built-in CIDR support).
// ---------------------------------------------------------------------------

// Returns [version, packed_bytes] or null for a malformed address.
function parse_ip(string $s) {
    $packed = @inet_pton($s);
    if ($packed === false) return null;
    return [strlen($packed) === 4 ? 4 : 6, $packed];
}

// Returns [version, packed_bytes, prefix] or null for a malformed CIDR.
function parse_cidr(string $cidr) {
    $slash = strrpos($cidr, "/");
    if ($slash === false) return null;
    $ip = parse_ip(substr($cidr, 0, $slash));
    if ($ip === null) return null;
    $prefix_str = substr($cidr, $slash + 1);
    if ($prefix_str === "" || !ctype_digit($prefix_str)) return null;
    $prefix = (int)$prefix_str;
    $bits = $ip[0] === 4 ? 32 : 128;
    if ($prefix < 0 || $prefix > $bits) return null;
    return [$ip[0], $ip[1], $prefix];
}

// Same-version membership by prefix bitmask over the packed bytes.
function ip_in_net(array $addr, array $net): bool {
    if ($addr[0] !== $net[0]) return false;
    $a = $addr[1];
    $n = $net[1];
    $prefix = $net[2];
    $full_bytes = intdiv($prefix, 8);
    $rem = $prefix % 8;
    if ($full_bytes > 0 && substr($a, 0, $full_bytes) !== substr($n, 0, $full_bytes)) return false;
    if ($rem !== 0) {
        $mask = (0xFF << (8 - $rem)) & 0xFF;
        if ((ord($a[$full_bytes]) & $mask) !== (ord($n[$full_bytes]) & $mask)) return false;
    }
    return true;
}

function ssrf_ok(string $url, ?string $resolved_ip, array $policy): bool {
    $ssrf = $policy["ssrf"];
    $u = parse_url($url);
    if ($u === false) return false;
    $scheme = isset($u["scheme"]) ? strtolower($u["scheme"]) : "";
    if (($ssrf["require_https"] ?? true) && $scheme !== "https") return false;
    if (isset($u["user"]) || isset($u["pass"])) return false; // userinfo -> reject
    $host = $u["host"] ?? "";
    if ($host === "") return false;
    $host = trim($host, "[]"); // strip IPv6 literal brackets

    $addr = parse_ip($host);
    if ($addr === null) {
        // hostname is a domain -> rely on the resolved address
        if ($resolved_ip === null || $resolved_ip === "") return false;
        $addr = parse_ip($resolved_ip);
        if ($addr === null) return false;
    }
    foreach ($ssrf["denied_cidrs"] as $cidr) {
        $net = parse_cidr($cidr);
        if ($net !== null && $addr[0] === $net[0] && ip_in_net($addr, $net)) return false;
    }
    return true;
}

// ---------------------------------------------------------------------------
// Rule 6: Ed25519 verify of the raw signing-base bytes under a raw 32-byte key.
// ---------------------------------------------------------------------------
function ed25519_verify(string $msg, string $sig, string $pk): bool {
    if (strlen($sig) !== 64 || strlen($pk) !== 32) return false;
    try {
        return sodium_crypto_sign_verify_detached($sig, $msg, $pk);
    } catch (\Throwable $e) {
        return false;
    }
}

// ---------------------------------------------------------------------------
// Driver.
// ---------------------------------------------------------------------------
$path = $argv[1] ?? DEFAULT_PATH;
$corpus = json_decode(file_get_contents($path), true);
$policy = $corpus["policy"];

$results = [];
$rec = function (string $section, string $note, bool $ok) use (&$results) {
    $results[] = [$section, $note, $ok];
};

// 1. wba_signing_base
foreach ($corpus["wba_signing_base"] as $c) {
    $decoded = base64_decode($c["signing_base_b64"]);
    // Build FIRST; `$c["ok"] && build(...)` would short-circuit negative cases
    // and never exercise the build-failure path.
    try {
        $built = build_signing_base($c["in"], $c["signature_params_raw"] ?? null);
        $ok = ($c["ok"] === true) && $built === $decoded;
    } catch (\Throwable $e) {
        $ok = !$c["ok"];
    }
    $rec("wba_signing_base", $c["note"], $ok);
}

// 2. wba_coverage
foreach ($corpus["wba_coverage"] as $c) {
    $rec("wba_coverage", $c["note"],
        coverage_ok($c["covered_components"], $policy) === $c["expect_accept"]);
}

// 3. wba_freshness
foreach ($corpus["wba_freshness"] as $c) {
    $rec("wba_freshness", $c["note"],
        freshness_ok($c["created"], $c["expires"], $c["now"], $policy) === $c["expect_accept"]);
}

// 4. wba_directory
foreach ($corpus["wba_directory"] as $c) {
    $rec("wba_directory", $c["note"],
        directory_ok($c["directory"], $c["keyid"]) === $c["expect_accept"]);
}

// 5. wba_directory_ssrf
foreach ($corpus["wba_directory_ssrf"] as $c) {
    $rec("wba_directory_ssrf", $c["note"],
        ssrf_ok($c["signature_agent_url"], $c["resolved_ip"] ?? null, $policy) === $c["expect_fetch_allowed"]);
}

// 6. wba_ed25519_verify
foreach ($corpus["wba_ed25519_verify"] as $c) {
    $base = base64_decode($c["signing_base_b64"]);
    $valid = ed25519_verify($base, hex2bin($c["sig_hex"]), hex2bin($c["pk_hex"]));
    $rec("wba_ed25519_verify", $c["note"], $valid === $c["expect_valid"]);
}

$fails = 0;
foreach ($results as [$section, $note, $ok]) {
    if (!$ok) {
        echo "FAIL  [$section] $note\n";
        $fails++;
    }
}
$total = count($results);
echo "\nphp (wba): " . ($total - $fails) . "/$total cases matched\n";
exit($fails === 0 ? 0 : 1);

File: runners/scala/verify_wba.scala
// Scala/JVM runner for the Web Bot Auth profile corpus (webbotauth_v0).
//
// Independently reproduces every verdict in the frozen corpus. The PROFILE rules
// (signing base, coverage completeness, freshness window, directory keyid
// resolution, Signature-Agent SSRF, Ed25519 verification) are re-implemented here
// from the corpus `policy` block, NOT imported from the generator's oracle, so a
// runner passing is genuine agreement rather than an echo. This mirrors the
// Python reference runner (runners/python/verify_wba.py), the independent KAT gate
// (tools/check_kat_wba.py) and the go/typescript runners case for case.
//
// JSON via Jackson; Ed25519 verify via Bouncy Castle; IPv4/IPv6 CIDR membership
// by byte-mask over java.net.InetAddress. Same `using` deps as the sibling runner
// (runners/scala/negative_v1.scala).
//
// Run:  scala-cli run --server=false verify_wba.scala -- <corpus path>
// Corpus path: argv[0], else the sibling repo default (runners/scala relative).
// Exit 0 iff every case matches, else 1.

//> using scala "3.8.4"
//> using dep "org.bouncycastle:bcprov-jdk18on:1.78.1"
//> using dep "com.fasterxml.jackson.core:jackson-databind:2.17.0"

import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

import java.io.File
import java.net.{InetAddress, URI}
import java.nio.charset.StandardCharsets
import java.util.Base64
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.*

class SigningBaseError(m: String) extends RuntimeException(m)

object VerifyWba:
  val DefaultPath = "../../corpus/webbotauth_v0/webbotauth_v0.json"

  // -------- generic JSON helpers --------

  // Null-safe string field access, mirroring dict.get(field) returning None.
  def txt(n: JsonNode, field: String): String =
    if n == null then null
    else
      val v = n.get(field)
      if v == null || v.isNull then null else v.asText()

  def note(c: JsonNode): String =
    if c.hasNonNull("note") then c.get("note").asText() else ""

  def hex(s: String): Array[Byte] =
    if s == null || (s.length % 2) != 0 then throw new NumberFormatException("bad hex")
    val out = new Array[Byte](s.length / 2)
    var i = 0
    while i < s.length do
      out(i / 2) = Integer.parseInt(s.substring(i, i + 2), 16).toByte
      i += 2
    out

  // ================= Rule 1: RFC 9421 Section 2.5 signing base =================
  // Hand-built by direct concatenation. Throws when a covered component's value
  // is missing (build impossible); the driver maps an unbuildable case to !ok.
  def reqStr(in: JsonNode, field: String, msg: String): String =
    val v = in.get(field)
    if v == null || v.isNull then throw new SigningBaseError(msg) else v.asText()

  def buildSigningBase(in: JsonNode, paramsRaw: String): String =
    val lines = ArrayBuffer[String]()
    val cc = in.get("covered_components")
    if cc == null || !cc.isArray then throw new SigningBaseError("no covered_components")
    cc.elements().asScala.foreach { compNode =>
      val name = compNode.asText().toLowerCase
      val value = name match
        case "@authority" => reqStr(in, "authority", "@authority covered but not supplied")
        case "@method"    => reqStr(in, "method", "@method covered but not supplied")
        case "@path"      => reqStr(in, "path", "@path covered but not supplied")
        case _ =>
          val h = in.get("headers")
          if h == null || !h.isObject || h.get(name) == null || h.get(name).isNull then
            throw new SigningBaseError(s"covered header $name not present")
          h.get(name).asText()
      lines += s"\"$name\": $value"
    }
    lines += s"\"@signature-params\": $paramsRaw"
    lines.mkString("\n")

  // ================= Rule 2: coverage completeness =================
  def coverageOk(covered: JsonNode, required: Seq[String]): Boolean =
    val have =
      if covered == null then Set.empty[String]
      else covered.elements().asScala.map(_.asText().toLowerCase).toSet
    required.forall(r => have.contains(r.toLowerCase))

  // ================= Rule 3: freshness window =================
  // All three must be JSON integers; expires > created; (created - skew) <= now < expires.
  def isIntNode(n: JsonNode): Boolean = n != null && n.isIntegralNumber

  def freshnessOk(c: JsonNode, skew: Long): Boolean =
    val cr = c.get("created"); val ex = c.get("expires"); val nw = c.get("now")
    if !(isIntNode(cr) && isIntNode(ex) && isIntNode(nw)) then return false
    val created = cr.asLong(); val expires = ex.asLong(); val now = nw.asLong()
    if expires <= created then return false
    (created - skew) <= now && now < expires

  // ================= Rule 4: directory keyid resolution =================
  def directoryOk(dir: JsonNode, keyid: String): Boolean =
    if dir == null then return false
    val it = dir.elements()
    while it.hasNext do
      val j = it.next()
      if txt(j, "kid") == keyid then
        val x = txt(j, "x")
        return txt(j, "kty") == "OKP" && txt(j, "crv") == "Ed25519" && x != null && x.nonEmpty
    false

  // ================= Rule 5: Signature-Agent SSRF guard =================
  // Strict IPv4 literal parse (no leading zeros, mirroring py3.9+ ipaddress).
  def parseV4(s: String): Array[Byte] =
    val parts = s.split("\\.", -1)
    if parts.length != 4 then return null
    val out = new Array[Byte](4)
    var i = 0
    while i < 4 do
      val p = parts(i)
      if p.isEmpty || p.length > 3 || !p.forall(_.isDigit) then return null
      if p.length > 1 && p(0) == '0' then return null // reject leading zeros
      val n = p.toInt
      if n > 255 then return null
      out(i) = n.toByte
      i += 1
    out

  // Parse an IP LITERAL to its network bytes (4 or 16). Never performs DNS:
  // strings containing ':' are treated as IPv6 literals (InetAddress parses them
  // numerically, never resolving); all others go through the strict IPv4 parser,
  // so a hostname returns null and the caller falls back to resolved_ip.
  def parseIp(s: String): Array[Byte] =
    if s == null || s.isEmpty then null
    else if s.contains(":") then
      try
        val a = InetAddress.getByName(s)
        val b = a.getAddress
        if b.length == 16 then b
        else
          // v4-mapped collapse (e.g. ::ffff:0:0): re-expand to 16 bytes so an
          // IPv6 literal always compares as IPv6.
          val b16 = new Array[Byte](16)
          b16(10) = 0xff.toByte; b16(11) = 0xff.toByte
          System.arraycopy(b, 0, b16, 12, 4)
          b16
      catch case _: Exception => null
    else parseV4(s)

  // Byte-mask CIDR membership. Same width (version) is required by the caller.
  def inNet(addr: Array[Byte], net: Array[Byte], prefix: Int): Boolean =
    if addr.length != net.length then return false
    val bits = addr.length * 8
    if prefix < 0 || prefix > bits then return false
    val fullBytes = prefix / 8
    val rem = prefix % 8
    var i = 0
    while i < fullBytes do
      if addr(i) != net(i) then return false
      i += 1
    if rem > 0 then
      val mask = (0xff << (8 - rem)) & 0xff
      if (addr(fullBytes) & mask) != (net(fullBytes) & mask) then return false
    true

  def ssrfOk(url: String, resolvedIp: String, requireHttps: Boolean, deniedCidrs: Seq[String]): Boolean =
    val uri = try new URI(url) catch case _: Exception => return false
    val scheme = Option(uri.getScheme).map(_.toLowerCase).getOrElse("")
    if requireHttps && scheme != "https" then return false
    val authority = uri.getRawAuthority
    if authority != null && authority.contains("@") then return false // userinfo -> reject
    var host = uri.getHost
    if host == null || host.isEmpty then return false
    if host.startsWith("[") && host.endsWith("]") then host = host.substring(1, host.length - 1)
    var addr = parseIp(host)
    if addr == null then
      // Not an IP literal: fall back to the resolved address.
      if resolvedIp == null || resolvedIp.isEmpty then return false
      addr = parseIp(resolvedIp)
      if addr == null then return false
    for cidr <- deniedCidrs do
      val slash = cidr.lastIndexOf('/')
      if slash > 0 then
        val netBytes = parseIp(cidr.substring(0, slash))
        val prefix = try cidr.substring(slash + 1).toInt catch case _: Exception => -1
        // Same IP version (byte width) only, mirroring the Python reference.
        if netBytes != null && prefix >= 0 && netBytes.length == addr.length && inNet(addr, netBytes, prefix) then
          return false
    true

  // ================= Rule 6: Ed25519 verify (Bouncy Castle) =================
  def ed25519Verify(msg: Array[Byte], sigHex: String, pkHex: String): Boolean =
    try
      val pk = hex(pkHex)
      if pk.length != 32 then false
      else
        val sig = hex(sigHex)
        val signer = new Ed25519Signer()
        signer.init(false, new Ed25519PublicKeyParameters(pk, 0))
        signer.update(msg, 0, msg.length)
        signer.verifySignature(sig)
    catch case _: Exception => false

  // ================= driver =================
  def main(args: Array[String]): Unit =
    val path = if args.nonEmpty then args(0) else DefaultPath
    val corpus = ObjectMapper().readTree(new File(path))
    val policy = corpus.get("policy")

    val required =
      val n = policy.get("required_covered_components")
      if n == null then Seq.empty[String] else n.elements().asScala.map(_.asText()).toSeq
    val skew =
      val n = policy.get("clock_skew_seconds")
      if n != null && n.isIntegralNumber then n.asLong() else 0L
    val ssrf = policy.get("ssrf")
    val requireHttps =
      val n = if ssrf == null then null else ssrf.get("require_https")
      if n != null && n.isBoolean then n.asBoolean() else true
    val deniedCidrs =
      val n = if ssrf == null then null else ssrf.get("denied_cidrs")
      if n == null then Seq.empty[String] else n.elements().asScala.map(_.asText()).toSeq

    val fails = ArrayBuffer[(String, String)]()
    var total = 0
    var matched = 0
    def rec(section: String, n: String, ok: Boolean): Unit =
      total += 1
      if ok then matched += 1 else fails += ((section, n))

    // 1. wba_signing_base
    corpus.get("wba_signing_base").elements().asScala.foreach { c =>
      val decoded = new String(Base64.getDecoder.decode(c.get("signing_base_b64").asText()), StandardCharsets.UTF_8)
      val okFlag = c.get("ok").asBoolean()
      val pr = { val p = txt(c, "signature_params_raw"); if p == null then "" else p }
      // Build FIRST; guarding on okFlag would skip the build-failure path on
      // negative cases.
      val ok = try
        val built = buildSigningBase(c.get("in"), pr)
        okFlag && built == decoded
      catch case _: Exception => !okFlag
      rec("wba_signing_base", note(c), ok)
    }

    // 2. wba_coverage
    corpus.get("wba_coverage").elements().asScala.foreach { c =>
      rec("wba_coverage", note(c), coverageOk(c.get("covered_components"), required) == c.get("expect_accept").asBoolean())
    }

    // 3. wba_freshness
    corpus.get("wba_freshness").elements().asScala.foreach { c =>
      rec("wba_freshness", note(c), freshnessOk(c, skew) == c.get("expect_accept").asBoolean())
    }

    // 4. wba_directory
    corpus.get("wba_directory").elements().asScala.foreach { c =>
      rec("wba_directory", note(c), directoryOk(c.get("directory"), c.get("keyid").asText()) == c.get("expect_accept").asBoolean())
    }

    // 5. wba_directory_ssrf
    corpus.get("wba_directory_ssrf").elements().asScala.foreach { c =>
      val resolved = txt(c, "resolved_ip")
      rec("wba_directory_ssrf", note(c),
        ssrfOk(c.get("signature_agent_url").asText(), resolved, requireHttps, deniedCidrs) == c.get("expect_fetch_allowed").asBoolean())
    }

    // 6. wba_ed25519_verify
    corpus.get("wba_ed25519_verify").elements().asScala.foreach { c =>
      val msg = Base64.getDecoder.decode(c.get("signing_base_b64").asText())
      val valid = ed25519Verify(msg, c.get("sig_hex").asText(), c.get("pk_hex").asText())
      rec("wba_ed25519_verify", note(c), valid == c.get("expect_valid").asBoolean())
    }

    fails.foreach { case (s, n) => println(s"FAIL  [$s] $n") }
    println(s"\nscala (wba): $matched/$total cases matched")
    System.exit(if matched == total then 0 else 1)
